spec/models/product_spec.rb

```diff
@@ -0,0 +1,87 @@
+require "rails_helper"
+require "active_storage_validations/matchers"
+
+RSpec.describe Product, type: :model do
+  describe "associations" do
+    it { should have_many(:comments).dependent(:destroy) }
+    it { should have_many(:cart_details).dependent(:destroy) }
+    it { should belong_to(:category).optional }
+    it { should have_one_attached(:image) }
+  end
+
+  describe "validations" do
+    context "when validations pass" do
+      let(:product) { build(:product) }
+
+      it "validates presence of name" do
+        expect(product).to validate_presence_of(:name)
+      end
+
+      it "validates presence of description" do
+        expect(product).to validate_presence_of(:description)
+      end
+
+      it "validates presence of price" do
+        expect(product).to validate_presence_of(:price)
+      end
+
+      it "validates numericality of price" do
+        expect(product).to validate_numericality_of(:price).is_greater_than(0)
+      end
+
+      it "validates presence of remain quantity" do
+        expect(product).to validate_presence_of(:remain_quantity)
+      end
+
+      it "validates numericality of remain quantity" do
+        expect(product).to validate_numericality_of(:remain_quantity).only_integer.is_greater_than_or_equal_to(0)
+      end
+
+      it "validates presence of category_id" do
+        expect(product).to validate_presence_of(:category_id)
+      end
+
+      it "validates image content type" do
+        expect(product).to validate_content_type_of(:image).allowing("image/png", "image/jpeg")
+      end
+
+      it "validates image size" do
+        expect(product).to validate_size_of(:image).less_than(Settings.max_image_data.megabytes)
+      end
+    end
+
+    context "when validations fail" do
```

Comment:
khi nào dùng describer, context, it

Reply:
Dạ,
describe: Dùng để nhóm các kiểm thử liên quan đến một đối tượng hoặc hành vi cụ thể.
context: Dùng để nhóm các kiểm thử dựa trên các điều kiện khác nhau.
it: Dùng để định nghĩa các kiểm thử cụ thể (example).